### PR TITLE
Provide a mechanism for getting project-level diagnostics

### PR DESCRIPTION
**Pull request by kevinushey**
This PR provides `.rs.lintProject()` as an (internal) entry-point for linting all R files within a package and displaying the source marker output. For example, here's what I see on linting `packrat`:

screen shot 2015-04-06 at 4 18 04 pm

I think this is the 'minimal' bit of work we should put in for project linting (we may want to wire up a button somewhere to perform this action as well; perhaps in the `Build` tools or something to that effect)

### CODE REVIEW

**Comment by kevinushey** [comment]
One blocker on this PR is the source document <-> file path map; currently running the function gives (in the error logs):

```
06 Apr 2015 23:13:19 [rsession-kevin] WARNING Failed to locate document with id ''; LOGGED FROM: void rstudio::session::modules::linter::<anonymous namespace>::addInferredSymbols(const rstudio::core::FilePath &, const std::string &, std::set<std::string> *) /Users/kevin/git/rstudio/src/cpp/session/modules/SessionLinter.cpp:105
06 Apr 2015 23:13:19 [rsession-kevin] ERROR system error 2 (No such file or directory); OCCURRED AT: rstudio::core::Error rstudio::session::source_database::get(const std::string &, boost::shared_ptr<SourceDocument>) /Users/kevin/git/rstudio/src/cpp/session/SessionSourceDatabase.cpp:518; LOGGED FROM: void rstudio::session::modules::linter::<anonymous namespace>::addInferredSymbols(const rstudio::core::FilePath &, const std::string &, std::set<std::string> *) /Users/kevin/git/rstudio/src/cpp/session/modules/SessionLinter.cpp:113
```

This will be alleviated by maintaining a mapping of file path -> document id.

**Comment by jjallaire** [comment]
Perhaps the Code menu could lose the "Show Diagnostics" command and have it replaced with (in it's own section):

Lint Source File
Lint Project

(not sure if those are the best command names)

**Comment by kevinushey** [comment]
I sort of feel like `Lint Project` doesn't really live there -- those operations are somewhat 'keyed' to the current document, whereas this is a project-wide action.

Here's a mockup of how it looks in the build pane -- the only thing 'unnatural' is that it doesn't actually initiate a typical 'build' action like the other operations; instead it just shows source markers. Any thoughts on that approach?

screen shot 2015-04-06 at 6 41 45 pm

**Comment by jjallaire** [comment]
The problem is that most projects don't even have a Build pane (only
packages do). Let's discuss realtime tomorrow.

On Apr 6, 2015, at 9:44 PM, Kevin Ushey <notifications@github.com> wrote:

I sort of feel like Lint Project doesn't really live there -- those
operations are somewhat 'keyed' to the current document, whereas this is a
project-wide action.

Here's a mockup of how it looks in the build pane -- the only thing
'unnatural' is that it doesn't actually initiate a typical 'build' action
like the other operations; instead it just shows source markers. Any
thoughts on that approach?

[image: screen shot 2015-04-06 at 6 41 45 pm]
<https://cloud.githubusercontent.com/assets/1976582/7015702/d81c65d8-dc8c-11e4-9976-9dcd6ba2f664.png>

—
Reply to this email directly or view it on GitHub
<https://github.com/rstudio/rstudio/pull/380#issuecomment-90308271>.

**Comment by jjallaire** [comment]
I don't think the Code menu is in any way restricted to the current source file (e.g. it has the Go to File/Function command which is project-wide, and the Forward/Back commands are also inter-file). In C/C++ mode it also includes "Find Usages" which is inter-file as well.

Pairing it with the command for the current file would also make it logical to include the command on the "Code Wizard" menu in the editor (thus making it a bit more discoverable).

**Comment by kevinushey** [comment]
Okay, I'm convinced -- I'll implement that.

**Comment by kevinushey** [comment]
Added now. TODO list:

- [ ] Implement `filepath` -> `docId` map; use it when necessary
- [ ] Use a progress indicator (as linting all files within a project may take a couple seconds)
- [ ] Figure out which files within the project should be eligible for linting

File: src/cpp/core/include/core/r_util/RTokenizer.hpp [hunk 1]
```diff
@@ -511,6 +511,28 @@ inline bool canContinueStatement(const RToken& rToken)
            canOpenArgumentList(rToken);
 }
 
+inline bool isSymbolNamed(const RToken& rToken,
+                          const std::wstring& name)
+{
+   // For strings, check if the content within the quotes
+   // is equal to the name provided. TODO: handle escaped
```
Comment (jjallaire) [inline src/cpp/core/include/core/r_util/RTokenizer.hpp]: Do we still need to handle the escaped quotes here?

File: src/cpp/session/modules/SessionCodeSearch.cpp [hunk 2]
```diff
@@ -919,6 +919,12 @@ void RSourceIndexes::update(
 
    // insert it
    indexes_[pDoc->id()] = pIndex;
+   
+   // cache a link between the document path and the id
+   std::string absPath = module_context::resolveAliasedPath(
```
Comment (jjallaire) [inline src/cpp/session/modules/SessionCodeSearch.cpp]: Should you also remove the entry from the cache when the document is removed?

File: src/cpp/core/include/core/r_util/RTokenizer.hpp (hunk 1 above)
Comment (kevinushey) [inline src/cpp/core/include/core/r_util/RTokenizer.hpp]: That note implies that we don't handle someone naming a symbol something like `"a\"function" <- function(x) { ... }`, which not a big deal IMO.

File: src/gwt/src/org/rstudio/studio/client/projects/Projects.java [hunk 3]
```diff
@@ -762,6 +762,28 @@ private void showOpenProjectDialog(
          onCompleted);  
    }
    
+   @Handler
+   public void onShowDiagnosticsProject()
+   {
+      final ProgressIndicator indicator = globalDisplay_.getProgressIndicator("Lint");
+      indicator.onProgress("Linting...");
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/projects/Projects.java]: I'd change this to "Analyzing project sources..."

File: src/cpp/session/modules/SessionCodeSearch.cpp (hunk 2 above)
Comment (kevinushey) [inline src/cpp/session/modules/SessionCodeSearch.cpp]: Yes, I should -- will add that.

File: src/gwt/src/org/rstudio/studio/client/workbench/commands/Commands.cmd.xml [hunk 4]
```diff
@@ -899,9 +901,12 @@ well as menu structures (for main menu and popup menus).
    <cmd id="reformatCode"
         menuLabel="Re_format Code"
         desc="Reformat the current line/selection"/>
-   <cmd id="lintActiveDocument"
+   <cmd id="showDiagnosticsActiveDocument"
         menuLabel="Show _Diagnostics"
         desc="Show diagnostics for the active document"/>
+   <cmd id="showDiagnosticsProject"
+        menuLabel="Show _Diagnostics (Project)"
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/commands/Commands.cmd.xml]: This needs a distinct mnemonic (_) character. The best way to figure out what's still available is to build the desktop version on Ubuntu then pull the Code menu down to see what characters are still left.

File: src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTargetWidget.java [hunk 5]
```diff
@@ -293,7 +293,8 @@ private Widget createCodeTransformMenuButton()
          menu.addSeparator();
          menu.addItem(commands_.reindent().createMenuItem(false));
          menu.addItem(commands_.reformatCode().createMenuItem(false));
-         menu.addItem(commands_.lintActiveDocument().createMenuItem(false));
```
Comment (jjallaire) [inline src/gwt/src/org/rstudio/studio/client/workbench/views/source/editors/text/TextEditingTargetWidget.java]: There should be a menu separator here

File: src/cpp/core/include/core/r_util/RTokenizer.hpp (hunk 1 above)
Comment (jjallaire) [inline src/cpp/core/include/core/r_util/RTokenizer.hpp]: Okay, fine re: not handling the quotes inside a quoted function.

**Pull request merged**